src/helpers.ts

```diff
@@ -657,3 +659,85 @@ export function snapshotStateAtSettlementRevenueWithdrawal(
+ * Creates a PurchaseDetails entity for a given token based on the provided event and returns its ID.
+ * The function processes events that are instances of GenArt721CoreV1Mint or GenArt721Core2PBABMint.
+ * If the event does not match these types, a warning is logged, and the function returns null.
+ *
+ * The function populates the PurchaseDetails with the minter's address, purchase currency address,
+ * and purchase currency symbol. It first checks for a minter configuration associated with the
+ * token's project. If found, this configuration is used. Otherwise, it falls back to the minter
+ * address from the event's transaction data, provided it's whitelisted in the project's contract.
+ *
+ * In cases where required data is missing or relevant configurations or contracts are not found,
+ * the function logs a warning and returns null.
+ *
+ * @param token - The token for which to create purchase details.
+ * @param project - The project associated with the token.
+ * @param event - The event triggering the creation of purchase details.
+ * @returns The ID of the newly created PurchaseDetails entity if successful, or null if not.
+ */
+export function createPrimaryPurchaseDetailsFromTokenMint<T>(
+  token: Token,
+  project: Project,
+  event: T
+): string | null {
+  if (
+    !(
+      event instanceof GenArt721CoreV1Mint ||
+      event instanceof GenArt721Core2PBABMint
+    )
+  ) {
+    log.warning(
+      "[WARN] createAndAssociatePurchaseDetails was called with an event that is not a GenArt721CoreV1Mint or GenArt721Core2PBABMint. The event was not processed.",
+      []
+    );
+    return null;
+  }
+
+  const purchaseDetails = new PrimaryPurchaseDetails(token.id);
+  purchaseDetails.token = token.id;
+  purchaseDetails.transactionHash = event.transaction.hash;
+
+  // If the token's project has an associated minter configuration, use it to populate the tokens purchase info
+  const minterConfigId = project.minterConfiguration;
+  if (minterConfigId) {
+    const minterConfig = ProjectMinterConfiguration.load(minterConfigId);
+    if (minterConfig) {
+      purchaseDetails.minterAddress = changetype<Bytes>(
+        ByteArray.fromHexString(minterConfig.minter)
+      );
+      purchaseDetails.currencyAddress = minterConfig.currencyAddress;
+      purchaseDetails.currencySymbol = minterConfig.currencySymbol;
+
+      purchaseDetails.save();
+      return purchaseDetails.id;
+    } else {
+      log.warning("Minter configuration with id {} does not exists", [
+        minterConfigId
+      ]);
+      return null;
+    }
+  } else {
+    // If no minter configuration is associated with the project, use the minter currently assigned to the contract
+    const contract = Contract.load(project.contract);
+    const to = changetype<Address>(event.transaction.to || Address.zero());
+
+    // If the transaction to address is a whitelisted minter, use it as the minter address
+    if (contract && contract.mintWhitelisted.includes(to)) {
```

Comment:
In general, assuming that `tx.to` is the minter address is a good assumption, although it is not foolproof (e.g. bot minting from contract, etc.). I don't think it is worth adding much more complexity here, but perhaps before falling back to `Address.zero()`, you could check if the length of `contract.mintWhitelisted` == 1, in which case you could assume the single item in the array was the minter.

Reply:
> In general, assuming that tx.to is the minter address is a good assumption, although it is not bulletproof (e.g. bot minting from contract, etc.)

I think the check for `contract.mintWhitelisted.includes(to)` should confirm the assumption though no?

Also, added the check for only a single allowed minter, good call!